### PR TITLE
Simplify migration workflow

### PR DESCRIPTION
**Pull request by oetherington**
- Remove schema hash checking
 - Remove migration order enforcement
 - Generate SQL schema as part of `yarn generate` (you should now run `yarn generate` after any change that changes the DB schema, such as adding an index or a postgres function)
 - Convert migration script to typescript (it's still a mess, but now it's a _well typed_ mess)
 - Automatically detect migration forum type from the file system (so you now just do `yarn migrate up dev` instead of needing `yarn migrate up dev ea` or `yarn migrate up dev lw`)
 - Fix `yarn migrate create` - you can now create a new stub migration with `yarn migrate create my_nice_new_migration`
 - Add a new CI job called `checkCodeGeneration` that fails if you forgot to run `yarn generate`

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1208509212185197) by [Unito](https://www.unito.io)

### CODE REVIEW

File: migrate.ts [hunk 1]
```diff
@@ -4,53 +4,76 @@
...
 
-const credentialsPath = (forumType) => {
+type ForumType = "lw" | "ea" | "none";
+
+const detectForumType = async (): Promise<ForumType> => {
+  const base = process.env.GITHUB_WORKSPACE ?? "..";
+  if (existsSync(`${base}/LessWrong-Credentials`)) {
+    return "lw";
+  }
+  if (existsSync(`${base}/ForumCredentials`)) {
```
Comment (b0b3rt) [inline migrate.ts:45]: Possibly the dirnames here should use `memorizedRepoNames` (which can be pulled out of the `credentialsPath` function)?

File: migrate.ts [hunk 2]
```diff
@@ -4,53 +4,76 @@
...
   const { getSqlClientOrThrow, setSqlClient } = require("./packages/lesswrong/server/sql/sqlClient");
   const { createSqlConnection } = require("./packages/lesswrong/server/sqlConnection");
   return { getSqlClientOrThrow, setSqlClient, createSqlConnection };
 }
 
-const credentialsPath = (forumType) => {
+type ForumType = "lw" | "ea" | "none";
+
+const detectForumType = async (): Promise<ForumType> => {
+  const base = process.env.GITHUB_WORKSPACE ?? "..";
```
Comment (b0b3rt) [inline migrate.ts:41]: Regrettably I suspect that not all LW devs have `LessWrong-Credentials` living in the same directory as `ForumMagnum`.  Am not really sure how to solve it in a convenient way... could just have this act as a forcing function, or just make everyone create an empty directory with that name (or a symlink?)

Comment (b0b3rt) [reply migrate.ts:41]: Oh, so it is.  For some reason I remembered this being an issue at some other point in time, but maybe I was remembering something else...

**Comment by oetherington** [comment]
@s-cheng Sorry - I only just realised this was never actually _assigned_ to you. No idea why.

Screenshot 2024-10-15 at 13 13 45

File: README.md [hunk 3]
```diff
@@ -78,10 +78,8 @@ psql forummagnum -f ./schema/accepted_schema.sql
 TODOs:
 
 * You won't have any database settings yet. TODO: add instructions.
-  * Start server and run it for ~30 sec which would create DB indexes 😅
```
Comment (s-cheng) [inline README.md:81]: Isn't this still the case?

**Review by s-cheng — APPROVED** [review]
I really appreciate this, thanks @oetherington! I didn't do a super thorough review of the code since I think we can fix issues later if they appear, and this is a big improvement.

File: README.md (hunk 3 above)
Comment (oetherington) [reply README.md:81]: Nope - see #9219 and #8307

**Pull request merged**